src/value-parser.ts

```diff
@@ -14,702 +15,764 @@ const PLP_NULL = Buffer.from([0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF]);
+    }
+
+    case 'NVarChar':
+    case 'NChar': {
+      // maxLength (unused?)
+      ({ offset } = readUInt16LE(buf, offset));
+
+      // collation (unsued?)
+      ({ offset } = readCollation(buf, offset));
+
+      return readNChars(buf, offset, dataLength);
+    }
+
+    default:
+      throw new Error('Invalid type!');
+  }
 }
 
-function readMaxChars(parser: Parser, codepage: string, callback: (value: unknown) => void) {
-  if (codepage == null) {
-    codepage = DEFAULT_ENCODING;
+function readBinary(buf: Buffer, offset: number, dataLength: number): Result<Buffer> {
+  if (buf.length < offset + dataLength) {
+    throw new NotEnoughDataError(offset + dataLength);
   }
 
-  readMax(parser, (data) => {
-    if (data) {
-      callback(iconv.decode(data, codepage));
-    } else {
-      callback(null);
-    }
-  });
+  return new Result(buf.slice(offset, offset + dataLength), offset + dataLength);
 }
 
-function readMaxNChars(parser: Parser, callback: (value: string | null) => void) {
-  readMax(parser, (data) => {
-    if (data) {
-      callback(data.toString('ucs2'));
-    } else {
-      callback(null);
-    }
-  });
+function readChars(buf: Buffer, offset: number, dataLength: number, codepage: string): Result<string> {
+  if (buf.length < offset + dataLength) {
+    throw new NotEnoughDataError(offset + dataLength);
+  }
+
+  return new Result(iconv.decode(buf.slice(offset, offset + dataLength), codepage ?? DEFAULT_ENCODING), offset + dataLength);
 }
 
-function readMax(parser: Parser, callback: (value: null | Buffer) => void) {
-  parser.readBuffer(8, (type) => {
-    if (type.equals(PLP_NULL)) {
-      return callback(null);
-    } else if (type.equals(UNKNOWN_PLP_LEN)) {
-      return readMaxUnknownLength(parser, callback);
-    } else {
-      const low = type.readUInt32LE(0);
-      const high = type.readUInt32LE(4);
-
-      if (high >= (2 << (53 - 32))) {
-        console.warn('Read UInt64LE > 53 bits : high=' + high + ', low=' + low);
-      }
+function readNChars(buf: Buffer, offset: number, dataLength: number): Result<string> {
+  if (buf.length < offset + dataLength) {
+    throw new NotEnoughDataError(offset + dataLength);
+  }
 
-      const expectedLength = low + (0x100000000 * high);
-      return readMaxKnownLength(parser, expectedLength, callback);
-    }
-  });
+  return new Result(buf.toString('ucs2', offset, offset + dataLength), offset + dataLength);
 }
 
-function readMaxKnownLength(parser: Parser, totalLength: number, callback: (value: null | Buffer) => void) {
-  const data = Buffer.alloc(totalLength, 0);
+async function readPLPStream(parser: Parser): Promise<null | Buffer[]> {
+  while (parser.buffer.length < parser.position + 8) {
+    await parser.waitForChunk();
+  }
+
+  const type = parser.buffer.slice(parser.position, parser.position + 8);
+  parser.position += 8;
+
+  if (type.equals(PLP_NULL)) {
+    return null;
+  }
 
+  const chunks: Buffer[] = [];
   let offset = 0;
```

Comment:
added  a offset var here for recording the read token length. We may able to handling it with some calculation with the paser.position but I think this maybe a bit more straight forward.

Reply:
Adding this variable to keep track how much we've read makes sense, but maybe we can call it something different than `offset`?

Reply:
how about call it readLength wich match the expectedLength

Reply:
changed it to currentLength instead